tools/labs/templates/debugging/locking2/locking2.c

```diff
@@ -1,39 +1,30 @@
 #include <linux/module.h>
 #include <linux/timer.h>
 
-spinlock_t lock;
-static struct timer_list timer;
- 
-static void foo(struct timer_list *unused)
-{
-	spin_lock(&lock);
+static DEFINE_SPINLOCK(lock);
 
-	spin_unlock(&lock);
+static void timerfn(struct timer_list *unused)
+{
+	pr_info("%s acquiring lock\n", __func__);
+	spin_lock(&lock);   pr_info("%s acquired lock\n", __func__);
+	spin_unlock(&lock); pr_info("%s released lock\n", __func__);
 }
 
-static int so2_locking_init(void)
+static DEFINE_TIMER(timer, timerfn);
+
+int init_module(void)
 {
-	pr_info("timer init\n");
-	timer_setup(&timer, foo, 0);
 	mod_timer(&timer, jiffies);
 
-	pr_info("timer setup ok!\n");
-
-	spin_lock_init(&lock);
-	spin_lock(&lock);
-
-	/* do work */
-	while(1);
-
+	pr_info("%s acquiring lock\n", __func__);
+	spin_lock(&lock);   pr_info("%s acquired lock\n", __func__);
+	spin_unlock(&lock); pr_info("%s released lock\n", __func__);
 	return 0;
 }
 
-static void so2_locking_exit(void)
+void exit_module(void)
 {
-	pr_info("locking exit\n");
 	del_timer_sync(&timer);
 }
 
 MODULE_LICENSE("GPL v2");
-module_init(so2_locking_init);
```

Comment:
Do we still need module_init and module_exit right?

Reply:
not if we use the default module entry/exit functions like we do here